course/src/matrix3.cpp

```diff
@@ -165,54 +165,35 @@ Matrix3 Matrix3::inverse() const {
   if (std::fabs(det) < 0.000001) {
     throw std::runtime_error("Matrix is non-invertible");
   }
-  const double& a = row_0_[0];
-  const double& b = row_0_[1];
-  const double& c = row_0_[2];
-  const double& d = row_1_[0];
-  const double& e = row_1_[1];
-  const double& f = row_1_[2];
-  const double& g = row_2_[0];
-  const double& h = row_2_[1];
-  const double& k = row_2_[2];
+  const double a = row_0_[0];
+  const double b = row_0_[1];
+  const double c = row_0_[2];
+  const double d = row_1_[0];
+  const double e = row_1_[1];
+  const double f = row_1_[2];
+  const double g = row_2_[0];
+  const double h = row_2_[1];
+  const double k = row_2_[2];
   return 1 / det *
          Matrix3((e * k - f * h), -(b * k - c * h), (b * f - c * e),
                  -(d * k - f * g), (a * k - c * g), -(a * f - c * d),
                  (d * h - e * g), -(a * h - b * h), (a * e - b * d));
 }
 
 Matrix3 Matrix3::product(const Matrix3& matrix) const {
-  const double& a11 = row_0_[0];
-  const double& a12 = row_0_[1];
-  const double& a13 = row_0_[2];
-  const double& a21 = row_1_[0];
-  const double& a22 = row_1_[1];
-  const double& a23 = row_1_[2];
-  const double& a31 = row_2_[0];
-  const double& a32 = row_2_[1];
-  const double& a33 = row_2_[2];
-  const double& b11 = matrix.row_0_[0];
-  const double& b12 = matrix.row_0_[1];
-  const double& b13 = matrix.row_0_[2];
-  const double& b21 = matrix.row_1_[0];
-  const double& b22 = matrix.row_1_[1];
-  const double& b23 = matrix.row_1_[2];
-  const double& b31 = matrix.row_2_[0];
-  const double& b32 = matrix.row_2_[1];
-  const double& b33 = matrix.row_2_[2];
-  return Matrix3{
-      a11 * b11 + a12 * b21 + a13 * b31, a11 * b12 + a12 * b22 + a13 * b32,
-      a11 * b23 + a12 * b23 + a13 * b33, a21 * b11 + a22 * b21 + a23 * b31,
-      a21 * b12 + a22 * b22 + a23 * b32, a21 * b13 + a22 * b23 + a23 * b33,
-      a31 * b11 + a32 * b21 + a33 * b31, a31 * b12 + a32 * b22 + a33 * b32,
-      a31 * b13 + a32 * b23 + a33 * b33};
+  Matrix3 aux = Matrix3{};
+  for (int i = 0; i < 3; ++i) {
+    for (int j = 0; j < 3; ++j) {
+      aux[i][j] = row(i).dot(matrix.col(i));
```

Comment:
This version is indeed correct, but is a bit slower than needed because of a subtlety: Notice that you're calling `row()` nine times to get only three different elements. Same goes with `col()`.

In both cases they will return by copy (because `this` is const here, so the const version of `row()` will be used, and `col()` always returns by copy). That means that you'll be creating an destroying at least 18 temporary `Vector3` objects, when you only needed 6. On top of that its also the cost of calling extra functions and any temporary object those functions create within.

That's some extra code that we can remove from that tight loop to make it faster. It's not a big deal in this code because the matrix is small, but if the matrix were 9x9 (a kalman filter, for instance), the ratio would be 162 calls made to 18 calls needed instead. 

You can improve this with an intermediate approach between this and the one you had used before:
```
r0 = row(0);
r1 = row(1);
r2 = row(2);
c0 = matrix.col(0);
c1 = matrix.col(1);
c2 = matrix.col(2);
return Matrix3{	
      r0.dot(c0), r0.dot(c1), r0.dot(c2),
      r1.dot(c0), r1.dot(c1), r1.dot(c2),
      r2.dot(c0), r2.dot(c1), r2.dot(c2),
};
```

Reply:
This is certainly a great explanation. Thanks a lot, Gera.